apps/backend/src/screenshot-diff/computeScreenshotDiff.ts

```diff
@@ -226,64 +226,72 @@ export const computeScreenshotDiff = async (
     };
   })();
 
-  await ScreenshotDiff.query()
-    .findById(screenshotDiff.id)
-    .patch({
-      ...diffData,
-      jobStatus: "complete",
-    });
-
   await Promise.all([
-    // Group similar diffs
-    groupSimilarDiffs({ diffFileKey: diffData.s3Id ?? null, buildId }),
     // Unlink images
     baseImage?.unlink(),
     compareImage.unlink(),
-    // screenshotDiff.testId
-    //   ? upsertTestStats({
-    //       testId: screenshotDiff.testId,
-    //       date: new Date(screenshotDiff.createdAt),
-    //       fileId: diffData.fileId ?? null,
-    //     })
-    //   : null,
   ]);
 
-  // Conclude build if it's the last diff
+  // Group similar diffs
+  if (diffData.s3Id) {
+    const lock = await getRedisLock();
+    await lock.acquire(["diff-group", diffData.s3Id], async () => {
+      await ScreenshotDiff.query().findById(screenshotDiff.id).patch(diffData);
+      await groupSimilarDiffs({ diffFileKey: diffData.s3Id, buildId });
+    });
+  }
+
+  // Insert stats
+  if (screenshotDiff.testId && build.type === "reference") {
+    await upsertTestStats({
+      testId: screenshotDiff.testId,
+      date: new Date(screenshotDiff.createdAt),
+      fileId: diffData.fileId ?? null,
+    });
+  }
+
+  await ScreenshotDiff.query()
+    .findById(screenshotDiff.id)
+    .patch({ ...diffData, jobStatus: "complete" });
+
```

Comment:
If the job dies here, the build is never concluded (because the jobStatus has been set to "complete" and it's not true, it's not complete.

But we use "jobStatus" in the concludeBuild, so... 

Anyway very few chances that we could have a problem here.